### ISSUE TITLE
Pylint does not smart enough when analyzes functions returning coroutine.

### ISSUE THREAD

**Issue by socketpair** [pylint-dev/pylint#9745]
### Bug description

```python
from asyncio import sleep
from typing import Awaitable

class A:
    async def fun(self) -> None:
        pass

class B(A)
    # Possibly -> Couroutine[Any, Any, None] would be better
    def fun(self) -> Awaitable[None]:
        return sleep(42)
```

### Command used

```shell
pylint a.py
```

### Pylint output

```python
W0236: Method 'fun' was expected to be 'async', found it instead as 'non-async' (invalid-overridden-method)
```

### Expected behavior

no warnings/errors

### Pylint version

```shell
pylint 3.0.4
astroid 3.0.3
Python 3.12.3 (main, Apr 17 2024, 00:00:00) [GCC 13.2.1 20240316 (Red Hat 13.2.1-7)]
```

### OS / Environment

Fedora 39

**Comment by DanielNoord** [comment]
I would argue that this is expected behaviour? We don't look at type annotations so I don't see how we could reliably do this without looking at whether `async` is missing. I also think this is a weird pattern you're introducing for which it is fine to have a somewhat opinionated linter check?

**Comment by socketpair** [comment]
It is used in existing code, and it is faster than

```python
async def fun():
    return await another_fun()
```

**Comment by DanielNoord** [comment]
Why is it faster? I have never heard about this, but if it is indeed we should turn this pattern itself into a recommendation?

**Comment by socketpair** [comment]
```python
from asyncio import run
from time import monotonic
from typing import Any, Coroutine


async def something() -> None:
    return


async def slow_fun() -> None:
    return await something()


def fast_fun() -> Coroutine[Any, Any, None]:
    return something()


async def test_slow() -> None:
    for i in range(1000_000):
        await slow_fun()


async def test_fast() -> None:
    for i in range(1000_000):
        await fast_fun()


run(something())  # warm-up

a = monotonic()
run(test_fast())
b = monotonic()
print('fast', b - a)

a = monotonic()
run(test_slow())
b = monotonic()
print('slow', b - a)
```

```
fast 0.0913987820094917
slow 0.12168889299209695
```

**Comment by DanielNoord** [comment]
Hmm that is interesting, I didn't know. I guess you're running into the overhead of `asyncio` itself here.

I guess we should indeed disable the warning in this case, although it still feels like a bit of an anti-pattern.
